### PR TITLE
Reworked Aggregator Bot

### PR DESCRIPTION
**Pull request by halaprix**

### CODE REVIEW

File: contracts/AutomationBotAggregator.sol [hunk 1]
```diff
@@ -52,6 +53,18 @@ contract AutomationBotAggregator {
...
+        uint256 cdpId,
+        bytes memory triggerData,
+        address commandAddress
+    ) private view returns (bytes32) {
+        bytes32 triggersHash = keccak256(
+            abi.encodePacked(cdpId, triggerData, serviceRegistry, commandAddress)
+        );
+
+        return triggersHash;
+    }
```
Comment (rkrasiuk) [inline contracts/AutomationBotAggregator.sol:66]:
```suggestion
    function getTriggerHash(
        uint256 cdpId,
        bytes memory triggerData,
        address commandAddress
    ) private view returns (bytes32) {
        return keccak256(
            abi.encodePacked(cdpId, triggerData, serviceRegistry, commandAddress)
        );
    }
```

File: contracts/AutomationBotAggregator.sol [hunk 2]
```diff
@@ -67,40 +80,43 @@ contract AutomationBotAggregator {
...
-
-        uint256 firstTriggerId = automationBot.triggersCounter() + 1;
+        uint256 firstTriggerId = bot.triggersCounter() + 1;
         uint256[] memory triggerIds = new uint256[](triggersData.length);
         for (uint256 i = 0; i < triggerTypes.length; i++) {
-            (bool status, ) = address(automationBot).delegatecall(
+            (bool status, ) = address(bot).delegatecall(
                 abi.encodeWithSelector(
-                    AutomationBot(automationBot).addTrigger.selector,
+                    AutomationBot(bot).addTrigger.selector,
```
Comment (rkrasiuk) [inline contracts/AutomationBotAggregator.sol:108]:
```suggestion
                    bot.addTrigger.selector,
```

File: contracts/AutomationBotAggregator.sol [hunk 3]
```diff
@@ -109,17 +125,13 @@ contract AutomationBotAggregator {
...
-
+        (AutomationBot bot, AutomationBotAggregator aggregator) = getBotAndAggregator();
         for (uint256 i = 0; i < triggerIds.length; i++) {
-            (bool status, ) = address(automationBot).delegatecall(
+            (bytes32 triggerHash, ) = bot.activeTriggers(triggerIds[i]);
+            require(groupId == aggregator.triggerGroup(triggerHash), "aggregator/invalid-group");
+            (bool status, ) = address(bot).delegatecall(
                 abi.encodeWithSelector(
-                    AutomationBot(automationBot).removeTrigger.selector,
+                    AutomationBot(bot).removeTrigger.selector,
```
Comment (rkrasiuk) [inline contracts/AutomationBotAggregator.sol:134]:
```suggestion
                    bot.removeTrigger.selector,
```

File: contracts/AutomationBotAggregator.sol [hunk 4]
```diff
@@ -128,82 +140,70 @@ contract AutomationBotAggregator {
...
-        );
 
-        (bool status, ) = address(automationBot).delegatecall(
+        require(aggregator.activeGroups(groupId) == cdpId, "aggregator/inactive-group");
+        require(aggregator.triggerGroup(newHash) == groupId, "aggregator/inactive-trigger");
+
+        (bool status, ) = address(bot).delegatecall(
             abi.encodeWithSelector(
-                AutomationBot(automationBot).addTrigger.selector,
+                AutomationBot(bot).addTrigger.selector,
```
Comment (rkrasiuk) [inline contracts/AutomationBotAggregator.sol:165]:
```suggestion
                bot.addTrigger.selector,
```

File: contracts/commands/CmBasicBuyCommand.sol [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+        return
+            nextCollRatio >= trigger.execCollRatio.wad() &&
+            nextPrice <= trigger.maxBuyPrice &&
+            trigger.targetCollRatio.wad().mul(currPrice).div(nextPrice) >
+            liquidationRatio.rayToWad() &&
+            baseFeeIsValid(trigger.maxBaseFeeInGwei);
+    }
+
+    function execute(
```
Comment (adamskrodzki) [inline contracts/commands/CmBasicBuyCommand.sol:113]: I believe instead of copy-pasting You can inherit and only override execute

**Review by adamskrodzki — CHANGES_REQUESTED** [review]

**Review by adamskrodzki — APPROVED** [review]

File: contracts/commands/CmBasicBuyCommand.sol (hunk 5 above)
Comment (adamskrodzki) [reply contracts/commands/CmBasicBuyCommand.sol:101]: https://github.com/OasisDEX/automation-smartcontracts/pull/109

**Review by adamskrodzki — CHANGES_REQUESTED** [review]
https://github.com/OasisDEX/automation-smartcontracts/pull/109

**Review by adamskrodzki — CHANGES_REQUESTED** [review]
Add deployment script for this specific feature in:
https://github.com/OasisDEX/automation-smartcontracts/tree/dev/scripts/deployment

**Comment by halaprix** [comment]
> Add deployment script for this specific feature in: https://github.com/OasisDEX/automation-smartcontracts/tree/dev/scripts/deployment

done : https://github.com/OasisDEX/automation-smartcontracts/pull/104/commits/4b1a3b16515af0024f1a18dea8a0a8042a54bba1

**Review by rkrasiuk — CHANGES_REQUESTED** [review]

File: contracts/commands/CmBaseMPACommand.sol [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+abstract contract CmBaseMPACommand is BaseMPACommand {
+    using RatioUtils for uint256;
+
+    struct TriggerData {
+        uint256 cdpId;
+        uint16 triggerType;
+        uint256 execCollRatio;
+        uint256 targetCollRatio;
+        uint256 BsPrice;
```
[the remark this answers is not in the document]
Comment (halaprix) [reply contracts/commands/CmBaseMPACommand.sol]: done

File: scripts/common/deploy-system.ts [hunk 7]
```diff
@@ -7,6 +7,8 @@ import {
     AutomationSwap,
     BasicBuyCommand,
     BasicSellCommand,
+    CmBasicBuyCommand,
```
[the remark this answers is not in the document]
Comment (halaprix) [reply scripts/common/deploy-system.ts:10]: https://github.com/OasisDEX/automation-smartcontracts/commit/4b1a3b16515af0024f1a18dea8a0a8042a54bba1

**Review by adamskrodzki — APPROVED** [review]

File: scripts/deployment/deploy-cm-basic-buy.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    console.log(`Network: ${network}`)
+
+    const system = await utils.getDefaultSystem()
+
+    system.cmBasicBuy = (await utils.deployContract(hre.ethers.getContractFactory('CmBasicBuyCommand'), [
+        utils.addresses.AUTOMATION_SERVICE_REGISTRY,
+    ])) as CmBasicBuyCommand
+    console.log(`CmBasicBuy Deployed: ${system.cmBasicBuy.address}`)
+
+    await configureRegistryEntries(utils, system, utils.addresses as AddressRegistry, [])
```
Comment (rkrasiuk) [inline scripts/deployment/deploy-cm-basic-buy.ts:20]: add command hash to `overwrites` parameter

File: scripts/deployment/deploy-cm-basic-sell.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    console.log(`Network: ${network}`)
+
+    const system = await utils.getDefaultSystem()
+
+    system.cmBasicSell = (await utils.deployContract(hre.ethers.getContractFactory('CmBasicSellCommand'), [
+        utils.addresses.AUTOMATION_SERVICE_REGISTRY,
+    ])) as CmBasicSellCommand
+    console.log(`CmBasicSell Deployed: ${system.cmBasicSell!.address}`)
+
+    await configureRegistryEntries(utils, system, utils.addresses as AddressRegistry, [])
```
Comment (rkrasiuk) [inline scripts/deployment/deploy-cm-basic-sell.ts:20]: same here

**Review by rkrasiuk — CHANGES_REQUESTED** [review]

**Review by adamskrodzki — APPROVED** [review]

**Pull request merged**